go/tasks/plugins/k8s/sagemaker/custom_training.go

```diff
@@ -0,0 +1,184 @@
+package sagemaker
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	trainingjobController "github.com/aws/amazon-sagemaker-operator-for-k8s/controllers/trainingjob"
+	"github.com/aws/aws-sdk-go/service/sagemaker"
+	flyteIdlCore "github.com/lyft/flyteidl/gen/pb-go/flyteidl/core"
+	taskError "github.com/lyft/flyteplugins/go/tasks/errors"
+	pluginsCore "github.com/lyft/flyteplugins/go/tasks/pluginmachinery/core"
+	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery/ioutils"
+	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery/k8s"
+	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery/utils"
+	"github.com/lyft/flytestdlib/logger"
+	"github.com/pkg/errors"
+
+	commonv1 "github.com/aws/amazon-sagemaker-operator-for-k8s/api/v1/common"
+
+	"github.com/lyft/flyteplugins/go/tasks/plugins/k8s/sagemaker/config"
+
+	trainingjobv1 "github.com/aws/amazon-sagemaker-operator-for-k8s/api/v1/trainingjob"
+	flyteSageMakerIdl "github.com/lyft/flyteidl/gen/pb-go/flyteidl/plugins/sagemaker"
+)
+
+func (m awsSagemakerPlugin) buildResourceForCustomTrainingJob(
+	ctx context.Context, taskCtx pluginsCore.TaskExecutionContext) (k8s.Resource, error) {
+
+	logger.Infof(ctx, "Building a training job resource for task [%v]", taskCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName())
+	taskTemplate, err := getTaskTemplate(ctx, taskCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the custom field of the task template back into the Hyperparameter Tuning Job struct generated in flyteidl
+	sagemakerTrainingJob := flyteSageMakerIdl.TrainingJob{}
+	err = utils.UnmarshalStruct(taskTemplate.GetCustom(), &sagemakerTrainingJob)
+	if err != nil {
+		return nil, errors.Wrapf(err, "invalid TrainingJob task specification: not able to unmarshal the custom field to [%s]", m.TaskType)
+	}
+
+	if sagemakerTrainingJob.GetAlgorithmSpecification() == nil {
+		return nil, errors.Errorf("The unmarshaled training job does not have a AlgorithmSpecification field")
+	}
+	if sagemakerTrainingJob.GetAlgorithmSpecification().GetAlgorithmName() != flyteSageMakerIdl.AlgorithmName_CUSTOM {
+		return nil, errors.Errorf("The algorithm name [%v] is not supported by the custom training job plugin",
+			sagemakerTrainingJob.GetAlgorithmSpecification().GetAlgorithmName().String())
+	}
+
+	inputChannels := make([]commonv1.Channel, 0)
+	inputModeString := strings.Title(strings.ToLower(sagemakerTrainingJob.GetAlgorithmSpecification().GetInputMode().String()))
+
+	jobName := taskCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName()
+	outputPath := taskCtx.OutputWriter().GetOutputPrefixPath().String()
+
+	if taskTemplate.GetContainer() == nil {
+		return nil, errors.Errorf("The task template points to a nil container")
```

Comment:
I think this should be permanent failure? an error here will cause only a system error right?